src/test/java/io/vertx/test/core/MetricsTest.java

```diff
@@ -270,28 +270,51 @@ private HandlerMetric assertRegistration(FakeEventBusMetrics metrics) {
 
   private void testHandlerProcessMessage(Vertx from, Vertx to, int expectedLocalCount) {
     FakeEventBusMetrics metrics = FakeMetricsBase.getMetrics(to.eventBus());
-    to.eventBus().consumer(ADDRESS1, msg -> {
-      HandlerMetric registration = assertRegistration(metrics);
-      assertEquals(ADDRESS1, registration.address);
-      assertEquals(null, registration.repliedAddress);
-      assertEquals(1, registration.beginCount.get());
-      assertEquals(0, registration.endCount.get());
-      assertEquals(0, registration.failureCount.get());
-      assertEquals(expectedLocalCount, registration.localCount.get());
-      msg.reply("pong");
-    }).completionHandler(onSuccess(v -> {
-      from.eventBus().send(ADDRESS1, "ping", reply -> {
+    CountDownLatch latch1 = new CountDownLatch(1);
+    CountDownLatch latch2 = new CountDownLatch(1);
+    to.runOnContext(v -> {
+      to.eventBus().consumer(ADDRESS1, msg -> {
         HandlerMetric registration = assertRegistration(metrics);
         assertEquals(ADDRESS1, registration.address);
         assertEquals(null, registration.repliedAddress);
+        assertEquals(1, registration.scheduleCount.get());
         assertEquals(1, registration.beginCount.get());
-        // This might take a little time
-        waitUntil(() -> 1 == registration.endCount.get());
+        assertEquals(0, registration.endCount.get());
         assertEquals(0, registration.failureCount.get());
         assertEquals(expectedLocalCount, registration.localCount.get());
-        testComplete();
-      });
-    }));
+        msg.reply("pong");
+      }).completionHandler(onSuccess(v2 -> {
+        to.runOnContext(v3 -> {
+          latch1.countDown();
+          try {
+            awaitLatch(latch2);
```

Comment:
Ha right. Got it, thanks!

Le 05/07/2016 à 17:17, Julien Viet a écrit :
> we need to wait for Latch2 because it guarantees that this particular
> event loop is busy when we send the message, so |from.eventBus().send|
> will schedule the message to send but it will not be delivered.